File: src/api/java/ArraySort.java
/**
Copyright (c) 2012-2014 Microsoft Corporation
   
Module Name:

    ArraySort.java

Abstract:

Author:

    @author Christoph Wintersteiger (cwinter) 2012-03-15

Notes:
    
**/

package com.microsoft.z3;

/**
 * Array sorts.
 **/
@SuppressWarnings("unchecked")
public class ArraySort<D extends Sort, R extends Sort> extends Sort
{
    /**
     * The domain of the array sort.
     * @throws Z3Exception 
     * @throws Z3Exception on error
     * @return a sort
     **/
    public D getDomain()
    {
        return (D) Sort.create(getContext(),
                Native.getArraySortDomain(getContext().nCtx(), getNativeObject()));
    }

    /**
     * The range of the array sort.
     * @throws Z3Exception 
     * @throws Z3Exception on error
     * @return a sort
     **/
    public R getRange()
    {
        return (R) Sort.create(getContext(),
                Native.getArraySortRange(getContext().nCtx(), getNativeObject()));
    }

    ArraySort(Context ctx, long obj)
    {
        super(ctx, obj);
    }

    ArraySort(Context ctx, D domain, R range)
    {
        super(ctx, Native.mkArraySort(ctx.nCtx(), domain.getNativeObject(),
                range.getNativeObject()));
    }

    ArraySort(Context ctx, Sort[] domains, R range)
    {
        super(ctx, Native.mkArraySortN(ctx.nCtx(), domains.length, AST.arrayToNative(domains),
                range.getNativeObject()));
    }
};

File: src/api/dotnet/ArraySort.cs
/*++
Copyright (c) 2012 Microsoft Corporation

Module Name:

    ArraySort.cs

Abstract:

    Z3 Managed API: Array Sorts

Author:

    Christoph Wintersteiger (cwinter) 2012-11-23

Notes:
    
--*/

using System.Diagnostics;
using System;

namespace Microsoft.Z3
{
    /// <summary>
    /// Array sorts.
    /// </summary>
    public class ArraySort : Sort
    {
        /// <summary>
        /// The domain of the array sort.
        /// </summary>
        public Sort Domain
        {
            get
            {

                return Sort.Create(Context, Native.Z3_get_array_sort_domain(Context.nCtx, NativeObject));
            }
        }

        /// <summary>
        /// The range of the array sort.
        /// </summary>
        public Sort Range
        {
            get
            {

                return Sort.Create(Context, Native.Z3_get_array_sort_range(Context.nCtx, NativeObject));
            }
        }

        #region Internal
        internal ArraySort(Context ctx, IntPtr obj) : base(ctx, obj) { Debug.Assert(ctx != null); }
        internal ArraySort(Context ctx, Sort domain, Sort range)
            : base(ctx, Native.Z3_mk_array_sort(ctx.nCtx, domain.NativeObject, range.NativeObject))
        {
            Debug.Assert(ctx != null);
            Debug.Assert(domain != null);
            Debug.Assert(range != null);
        }
        internal ArraySort(Context ctx, Sort[] domain, Sort range)
            : base(ctx, Native.Z3_mk_array_sort_n(ctx.nCtx, (uint)domain.Length, AST.ArrayToNative(domain), range.NativeObject))
        {
            Debug.Assert(ctx != null);
            Debug.Assert(domain != null);
            Debug.Assert(range != null);
        }
        #endregion
    };

}
